packages/tpetra/core/example/BlockCrs/Tpetra_TestBlockCrs.cpp

```diff
@@ -0,0 +1,526 @@
+    if (std::is_same<exec_space,Kokkos::Serial>::value)
+      std::cout << "Kokkos::Serial " << std::endl;
+    else
+      exec_space::print_configuration(std::cout, false);
+  }
+
+  MeshDatabase mesh(comm,
+                    num_global_elements_i,
+                    num_global_elements_j,
+                    num_global_elements_k,
+                    num_procs_i,
+                    num_procs_j,
+                    num_procs_k);
+
+  const auto sb = mesh.getStructuredBlock();
+  const auto part = mesh.getStructuredBlockPart();
+
+  const auto num_owned_elements = mesh.getNumOwnedElements();
+  const auto num_remote_elements = mesh.getNumRemoteElements();
+  const auto num_owned_and_remote_elements = mesh.getNumElements();
+
+#if PRINT_VERBOSE
+  mesh.print(std::cout);
+#endif
+  
+  {
+    TimeMonitor timerGlobal(*TimeMonitor::getNewTimer("X) Global"));
+    
+    // Build Tpetra Maps
+    // -----------------
+    
+    const Tpetra::global_size_t TpetraComputeGlobalNumElements 
+      = Teuchos::OrdinalTraits<Tpetra::global_size_t>::invalid();  
+    
+    // - all internal views are allocated on device; mirror as mesh database is constructed on host  
+    const auto mesh_gids_host = mesh.getElementGlobalIDs();
+    const auto mesh_gids = Kokkos::create_mirror_view(typename exec_space::memory_space(), mesh.getElementGlobalIDs());
+    Kokkos::deep_copy(mesh_gids, mesh_gids_host);
+
+    // for convenience, separate the access to owned and remote gids
+    const auto owned_gids  = Kokkos::subview(mesh_gids, local_ordinal_range_type(                 0, num_owned_elements));
+    const auto remote_gids = Kokkos::subview(mesh_gids, local_ordinal_range_type(num_owned_elements, num_owned_and_remote_elements));
+    
+    RCP<const map_type> row_map = rcp(new map_type(TpetraComputeGlobalNumElements, owned_gids, 0, comm));
+    RCP<const map_type> col_map = rcp(new map_type(TpetraComputeGlobalNumElements, mesh_gids, 0, comm));
+    
+#if PRINT_VERBOSE
+    row_map->describe(*out);
+    col_map->describe(*out);
+#endif
+    
+    // Graph Construction
+    // ------------------
+    // local graph is constructed on device space
+    typedef tpetra_crs_graph_type::local_graph_type local_graph_type;
+    typedef local_graph_type::row_map_type::non_const_type rowptr_view_type;
+    typedef typename local_graph_type::entries_type colidx_view_type;
+
+    rowptr_view_type rowptr;
+    colidx_view_type colidx;
+    {
+      TimeMonitor timerLocalGraphConstruction(*TimeMonitor::getNewTimer("0) LocalGraphConstruction"));
+      
+      rowptr = rowptr_view_type("rowptr", num_owned_elements + 1);
+
+      local_ordinal_range_type owned_range_i, owned_range_j, owned_range_k;
+      local_ordinal_range_type remote_range_i, remote_range_j, remote_range_k;
+
+      part.getOwnedRange(owned_range_i, owned_range_j, owned_range_k);
+      part.getRemoteRange(sb, remote_range_i, remote_range_j, remote_range_k);
+      
+      // count # of nonzeros per row
```

Comment:
Please break this `parallel_scan` out into a separate function so that this is easier to read.

Reply:
done.